src/faucet/Faucet.sol

```diff
@@ -0,0 +1,90 @@
+/*HOLOGRAPH_LICENSE_HEADER*/
+
+/*SOLIDITY_COMPILER_VERSION*/
+
+import "../abstract/Initializable.sol";
+import "../interface/ERC20Holograph.sol";
+import "../interface/IInitializable.sol";
+
+contract Faucet is Initializable {
+  address public owner;
+  ERC20Holograph public token;
+
+  uint256 public faucetDripAmount = 100 ether;
+  uint256 public faucetCooldown = 24 hours;
+
+  mapping(address => uint256) lastAccessTime;
+
+  /**
+  * @notice Constructor is empty and not utilised.
+  * @dev To make exact CREATE2 deployment possible, constructor is left empty. We utilize the "init" function instead.
+  */
+  constructor() {}
+
+  /**
+  * @notice Initializes the token.
+  * @dev Special function to allow a one time initialisation on deployment.
+  */
+  function init(bytes memory data) external override returns (bytes4) {
+    require(!_isInitialized(), "Faucet already initialized");
+    (address contractOwner_, address tokenInstance_) = abi.decode(data, (address, address));
+    token = ERC20Holograph(tokenInstance_);
+    owner = contractOwner_;
+    _setInitialized();
+    return IInitializable.init.selector;
+  }
+
+  /// @notice Get tokens from faucet's own balance. Rate limited.
+  function requestTokens() external {
+    require(isAllowedToWithdraw(msg.sender), 'Come back later');
+    require(token.balanceOf(address(this)) >= faucetDripAmount, "Faucet is empty");
+    lastAccessTime[msg.sender] = block.timestamp;
+    token.transfer(msg.sender, faucetDripAmount);
+  }
+
+  /// @notice Grant tokens to receiver from faucet's own balance. Not rate limited.
+  function grantTokens(address address_) external onlyOwner {
```

Comment:
This is a nit, but it's more typical for function variables to use the underscore at the front of the variable name in Solidity

Reply:
Yeah, you can see it in our repo too

```
contract CxipERC721 is ERC721H {
  /**
   * @dev Internal reference used for minting incremental token ids.
   */
  uint224 private _currentTokenId;

  /**
   * @dev Enum of type of token URI to use globally for the entire contract.
   */
  TokenUriType private _uriType;

  /**
   * @dev Enum mapping of type of token URI to use for specific tokenId.
   */
  mapping(uint256 => TokenUriType) private _tokenUriType;

  /**
   * @dev Mapping of IPFS URIs for tokenIds.
   */
  mapping(uint256 => mapping(TokenUriType => string)) private _tokenURIs;

  /**
   * @notice Constructor is empty and not utilised.
   * @dev To make exact CREATE2 deployment possible, constructor is left empty. We utilize the "init" function instead.
   */
  constructor() {}

  /**
   * @notice Initializes the collection.
   * @dev Special function to allow a one time initialisation on deployment. Also configures and deploys royalties.
   */
  function init(bytes memory data) external override returns (bytes4) {
    // we set this as default type since that's what Mint is currently using
    _uriType = TokenUriType.IPFS;
    address owner = abi.decode(data, (address));
    _owner = owner;
    // run underlying initializer logic
    return _init(data);
  }

  /**
   * @notice Get's the URI of the token.
   * @return string The URI.
   */
  function tokenURI(uint256 _tokenId) external view onlyHolographer returns (string memory) {
    TokenUriType uriType = _tokenUriType[_tokenId];
    if (uriType == TokenUriType.UNDEFINED) {
      uriType = _uriType;
    }
    return
      string(
        abi.encodePacked(
          IInterfaces(IHolograph(IHolographer(holographer()).getHolograph()).getInterfaces()).getUriPrepend(uriType),
          _tokenURIs[_tokenId][uriType]
        )
      );
  }
```

IMO this is very confusing and my approach is better.

Reply:
I do like the idea of using underscores at end of variable name to define that it came from external data. Visually easy to understand this way where the data came from. Underscore in front for internal private variables, underscore at end to show user submitted var, no underscores means locally defined var (inside function). Could be interesting to try.

Reply:
My opinion is to just stick with underscores in the front for now for consistency, but it's not a big deal either way